components/gantry/oneaxis/oneaxis.go

```diff
@@ -119,6 +121,18 @@ func newOneAxis(ctx context.Context, deps resource.Dependencies, conf resource.C
 func (g *oneAxis) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
 	g.mu.Lock()
 	defer g.mu.Unlock()
+
+	if g.motor != nil {
+		if err := g.motor.Stop(ctx, nil); err != nil {
+			return err
+		}
+	}
+
+	if g.cancelFunc != nil {
+		g.cancelFunc()
+		g.activeBackgroundWorkers.Wait()
+	}
+
```

Comment:
okay I'm having a new problem here with the tests, and assuming the same issue will happen in practice. the test hangs on line 133, it is waiting for background workers to be done, but I guess there's nothing in the `checkHit` function saying that it should be done? I thought that's what `g.cancelFunc` was supposed to do but it doesn't seem to. any thoughts @penguinland

Reply:
try injecting a pin that eventually returns a hit in the tests?

Reply:
actually, it already returns a true `Get` bool almost immediately, when you're debugging, where is it hanging? Also try  adding an `if ctx.Err != nil {}` statement in that for loop.

Reply:
@randhid it's hanging on line 133. @penguinland I added those lock and unlocks because there was a race condition error in the tests. I can try moving around the locks but I assume the race condition will still be an issue if I remove it completely. I also tried moving the lock/unlock in reconfigure to after the cancelfunc but it still hangs.

Reply:
Martha and I figured it out: it's not a deadlock, it's a _livelock!_ This was in a test which injected a fake limit switch that always reported that it was tripped. The main coroutine is stuck waiting for the background coroutine to exit, and the background coroutine was backing off the motor until the limit switch was no longer tripped, without ever checking if it was supposed to shut down. I like the behavior of the background coroutine here: if it's in the middle of rescuing the gantry, that shouldn't be abandoned. I think we should change the test's injected pin to sometimes say it's tripped, and sometimes say it's not tripped so the coroutine can finish backing off.